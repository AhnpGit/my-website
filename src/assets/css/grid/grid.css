@lost gutter 32px;
@lost flexbox no-flex;
@lost cycle auto;
@lost clearing both;
@lost rounder 99.9;
@lost --beta-direction ltr;

.grid-item {
  background: #2c3e50;
  margin-bottom: 20px;
  color: #FFF;
  height: 100px;
}

.grid-one {
  lost-center: 100% 20px;
}

@media screen and (min-width: 320px) {
  .grid-item {
    lost-column: 1/1;
  }
}

@media screen and (min-width: 768px) {
  .grid-item {
    lost-column: 1/3;
  }
}

@media screen and (min-width: 1024px) {
  .grid-item {
    lost-column: 1/4 4;
  }
}
