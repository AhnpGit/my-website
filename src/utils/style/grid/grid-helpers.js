// @flow

import addPx from 'add-px-to-style'
import hyphenate from 'hyphenate-style-name'
import { css } from 'styled-components'

/* calculate column width */
export const calcValue = (fraction: string, gutter: string = '0', rounder: number = 99.9, unit: string = '%') => {
  let gutterLogic = ''

  if (gutter !== '0') {
    gutterLogic = ` - (${gutter} - ${gutter} * ${fraction})`
  }

  let calcValue = `calc((${rounder}${unit} * ${fraction})${gutterLogic})`
  return calcValue
}

/** CREATE CSS BLOCKS
 *
 * This function is used to create (css-ish)style blocks which can be inserted into styled-components. This function is
 * far from complete, but for now will serve our purposes.
 *
 * @param {object} obj     - Object that holds the styling. The key is a css-property, the value is the css-value
 *
 * @param {Array} [addToObj] - Array that contains code to be added inside the obj code block. For now this is only used
 *                           to insert nth-child styling. In the future we could expand this to add even other nested
 *                           styles to the obj.
 *                           Index[0] of addToObj contains the css-selector, the other indices contain the desired
 *                           styles. This option isn't fully functional yet and should be improved in future iterations.
 *
 * @returns {function} css - The function returns the { css } function from styled-components which creates a template
 *                           literal that composes the css. It gets its input from the results of the operations on
 *                           the obj and addToObj attributes.
 */

export const newBlock = (obj: { [key: string]: ?any }, addToObj?: Array<any>) => {
  const keys = Object.keys(obj)

  /* build child node css */
  let childNode
  if (addToObj !== undefined) {
    let selector: string = addToObj[0]
    if (selector.indexOf('&:') !== -1) {
      childNode = `${selector} { 
        margin-left: auto;
        margin-right: 0;
      }`
    }
  }

  /* iterate through the keys that were obtained from the obj attribute. Use these keys and their values to create
   * css properties and their values. */
  let i, len = keys.length
  let result = ''
  for (i = 0 ; i < len ; i++) {
    let key = keys[i]
    let val = obj[key]
    result += hyphenate(key) + ':' + addPx(key, val) + ';'
  }

  /* combine the output of both the obj and the addToObj attributes into one code block so that we can insert them into
   * our styled-components. */
  return css`
    ${result}
    ${childNode}
  `
}
