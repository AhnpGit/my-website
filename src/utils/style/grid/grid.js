// @flow

import { css } from 'styled-components'
import { calcValue, newBlock } from './grid-helpers'

// TODO Create fallback for IE9 & 10 in this script

export const settings = {
  gutter: '32px',
  flexbox: 'no-flex',
  cycle: 'auto',
  clearing: 'both',
  rounder: 99.9,
  unit: '%',
  direction: 'ltr',
}

// TODO: Maybe these types should be moved to a separate file, containing all Flex or maybe even all CSS related types.
type FlexDirection = 'row' | 'row-reverse' | 'column' | 'column-reverse'
type FlexWrap = 'nowrap' | 'wrap' | 'wrap-reverse'

/* Grid Wrapper */
export const gridWrapper = (direction: FlexDirection = 'row', wrap: FlexWrap = 'wrap', padding: ?string = null) => {
  const flexFlow = direction + ' ' + wrap
  if (!padding) {
    return css`
      display: flex;
      flex-flow: ${flexFlow};
    `
  } else {
    return css`
      display: flex;
      flex-flow: ${flexFlow};
      padding-left: ${settings.gutter};
      padding-right: ${settings.gutter};
    `
  }
}


/** GRID COLUMN
 *
 * @param {string} decl           - The function accepts a string of attributes that it uses to set a uniform grid. It
 *                                  accepts the following parameters:
 *                                - fraction (e.g. 1/3 to give each column a width that is a third of the parent minus
 *                                  gutters. The function makes sure that if the number of items in the grid is bigger
 *                                  than the denominator (the last number in the fraction), the last item in the row
 *                                  has no margin to the right.
 *                                - [gutter] A pixel value can be used if the gutter for the pixel should be different
 *                                  from the default value in settings.gutter.
 *                                - [flex/no-flex] If the display property of the column item needs to be different from
 *                                  the default display value in settings.flexbox. If no-flex is chosen the item will
 *                                  be a block-level item.
 *
 * @returns {Function} newBlock   - A function that puts all the necessary styles inside the newBlock function and
 *                                  makes sure that the returned css gets inserted into the styled components.
 */

export const gridColumn = (decl: string) => {
  let declArray = decl.split(' ')
  let gridColumn = declArray[0]
  let columnCycle: string
  let columnGutter = settings.gutter
  let columnFlexbox = settings.flexbox
  let columnRounder = settings.rounder
  let unit = settings.unit
  let declaration = {}

  /* Determin if we can use the default cycle setting or if we should */
  if (settings.cycle === 'auto') {
    columnCycle = gridColumn.split('/')[1]
  } else {
    columnCycle = settings.cycle
  }

  if (declArray[1] === 'flex' || declArray[1] === 'no-flex' || declArray[1] === 'auto') {
    columnCycle = declArray[0].split('/')[1]
  }

  if (declArray[1] !== undefined && declArray[1].search(/px/) !== -1) {
    columnGutter = declArray[1]
  }

  if (declArray[2] !== undefined && declArray[2].search(/px/) !== -1) {
    columnGutter = declArray[2]
  }

  if (declArray.includes('flex')) {
    columnFlexbox = 'flex'
  }

  if (declArray.includes('no-flex')) {
    columnFlexbox = 'no-flex'
  }

  /* build generic styles */
  declaration['margin-left'] = '0'
  declaration['margin-right'] = columnGutter
  declaration['margin-bottom'] = columnGutter
  declaration['flex-grow'] = '0'
  declaration['flex-shrink'] = '0'
  declaration['flex-basis'] = calcValue(gridColumn, columnGutter, columnRounder, unit)
  /* flex columns */
  if (columnFlexbox === 'flex') {
    declaration['display'] = 'flex'
    if (columnCycle !== 0) {
      return newBlock(
        declaration, [
          `&:nth-child(${columnCycle}n)`,
          ['margin-right', 'margin-left'],
          ['0', 'auto']]
      )
    } else {
      return newBlock(declaration)
    }
    /* block columns*/
  } else {
    if (columnCycle !== 0) {
      return newBlock(
        declaration, [
          `&:nth-child(${columnCycle}n)`,
          ['margin-right', 'margin-left'],
          ['0', 'auto']]
      )
    } else {
      return newBlock(declaration)
    }
  }
}
