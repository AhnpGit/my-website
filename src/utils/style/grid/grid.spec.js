import { gridWrapper, gridColumn } from './grid'

describe('grid', () => {

  describe('gridWrapper', () => {
    it('by default returns a tagged template literal that sets default styling for a GridWrapper element', () => {
      const result = gridWrapper()
      expect(result).toMatchSnapshot()
    })
    it('sets the direction to column when column is passed as first argument', () => {
      const result = gridWrapper('column', 'wrap')
      expect(result).toMatchSnapshot()
    })
    it('turns of wrap when nowrap is passed as second argument', () => {
      const result = gridWrapper('row', 'nowrap')
      expect(result).toMatchSnapshot()
    })
    it('sets a left and right padding when a pixel value is provided as a third argument', () => {
      const result = gridWrapper('row', 'wrap', '16px')
      expect(result).toMatchSnapshot()
    })
  })

  describe('gridColumn', () => {
    it('sets a width that is the same as the fraction passed as the first argument', () => {
      const result = gridColumn('1/3')
      expect(result).toMatchSnapshot()
    })
    it('sets a gutter when the default value is overwritten by the invocation of the function', () => {
      const result = gridColumn('1/4 16px')
      expect(result).toMatchSnapshot()
    })
    it('gives the the element a display of flex when flex is passed as the third argument', () => {
      const result = gridColumn('1/2 32px flex')
      expect(result).toMatchSnapshot()
    })
    it('gives the the element a display of noflex when noflex is passed as the third argument', () => {
      const result = gridColumn('1/2 32px noflex')
      expect(result).toMatchSnapshot()
    })
  })

})
