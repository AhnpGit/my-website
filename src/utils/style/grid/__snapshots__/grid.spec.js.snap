// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`grid gridColumn gives the the element a display of flex when flex is passed as the third argument 1`] = `
Array [
  "
    ",
  "margin-left:0;margin-right:32px;margin-bottom:32px;flex-grow:0;flex-shrink:0;flex-basis:calc((99.9% * 1/2) - (32px - 32px * 1/2));display:flex;",
  "
    ",
  "&:nth-child(2n) { 
        margin-left: auto;
        margin-right: 0;
      }",
  "
  ",
]
`;

exports[`grid gridColumn gives the the element a display of noflex when noflex is passed as the third argument 1`] = `
Array [
  "
    ",
  "margin-left:0;margin-right:32px;margin-bottom:32px;flex-grow:0;flex-shrink:0;flex-basis:calc((99.9% * 1/2) - (32px - 32px * 1/2));",
  "
    ",
  "&:nth-child(2n) { 
        margin-left: auto;
        margin-right: 0;
      }",
  "
  ",
]
`;

exports[`grid gridColumn sets a gutter when the default value is overwritten by the invocation of the function 1`] = `
Array [
  "
    ",
  "margin-left:0;margin-right:16px;margin-bottom:16px;flex-grow:0;flex-shrink:0;flex-basis:calc((99.9% * 1/4) - (16px - 16px * 1/4));",
  "
    ",
  "&:nth-child(4n) { 
        margin-left: auto;
        margin-right: 0;
      }",
  "
  ",
]
`;

exports[`grid gridColumn sets a width that is the same as the fraction passed as the first argument 1`] = `
Array [
  "
    ",
  "margin-left:0;margin-right:32px;margin-bottom:32px;flex-grow:0;flex-shrink:0;flex-basis:calc((99.9% * 1/3) - (32px - 32px * 1/3));",
  "
    ",
  "&:nth-child(3n) { 
        margin-left: auto;
        margin-right: 0;
      }",
  "
  ",
]
`;

exports[`grid gridWrapper by default returns a tagged template literal that sets default styling for a GridWrapper element 1`] = `
Array [
  "
      display: flex;
      flex-flow: ",
  "row wrap",
  ";
    ",
]
`;

exports[`grid gridWrapper sets a left and right padding when a pixel value is provided as a third argument 1`] = `
Array [
  "
      display: flex;
      flex-flow: ",
  "row wrap",
  ";
      padding-left: ",
  "32px",
  ";
      padding-right: ",
  "32px",
  ";
    ",
]
`;

exports[`grid gridWrapper sets the direction to column when column is passed as first argument 1`] = `
Array [
  "
      display: flex;
      flex-flow: ",
  "column wrap",
  ";
    ",
]
`;

exports[`grid gridWrapper turns of wrap when nowrap is passed as second argument 1`] = `
Array [
  "
      display: flex;
      flex-flow: ",
  "row nowrap",
  ";
    ",
]
`;
